src/plaid/plaid.service.ts

```diff
@@ -0,0 +1,79 @@
+import { Injectable } from '@nestjs/common'
+import { ModelId } from '../core/model'
+import { EncryptionService } from '../encryption/encryption.service'
+import { PlaidItem, tableName } from './plaidItem.entity'
+import { InjectRepository } from '@nestjs/typeorm'
+import { Repository } from 'typeorm'
+import plaid, { Account, Client } from 'plaid'
+import { ConfigService } from '@nestjs/config'
+import { IPlaidConfig } from '../config/plaidConfig'
+
+@Injectable()
+export class PlaidService {
+  private readonly client: Client
+  constructor(
+    private readonly encryptionService: EncryptionService,
+    private readonly configService: ConfigService,
+    @InjectRepository(PlaidItem) private readonly plaidSessionRepo: Repository<PlaidItem>,
+  ) {
+    const config = this.configService.get<IPlaidConfig>('plaid')
+    if (!config) {
+      throw new Error('Missing Plaid Config')
+    }
+    this.client = new plaid.Client(
+      config.clientId,
+      config.secret,
+      config.publicKey,
+      plaid.environments[config.env],
+      { version: '2019-05-29', clientApp: 'Bitmatica' },
+    )
+  }
+
+  async exchangePublicToken(userId: ModelId, publicToken: string): Promise<string> {
+    const response = await this.client.exchangePublicToken(publicToken)
+    const item = new PlaidItem()
+    item.userId = userId
+    item.itemId = response.item_id
+    item.accessToken = await this.encryptionService.encrypt(
+      response.access_token,
+      PlaidService.encryptionContext(userId, response.item_id),
+    )
+    await this.plaidSessionRepo.save(item)
+    return item.itemId
+  }
+
+  async getPlaidAccounts(userId: ModelId, itemId: string): Promise<Array<Account>> {
+    const accessToken = await this.getAccessToken(userId, itemId)
+    if (!accessToken) {
+      return Promise.reject('Access token invalid')
+    }
+
+    return (await this.client.getAccounts(accessToken)).accounts
+  }
+
+  async getAccessToken(userId: ModelId, itemId: string): Promise<string | undefined> {
+    // TODO: cache accessToken in-memory to avoid decryption calls?
+    const item = await this.plaidSessionRepo.findOne({ userId, itemId })
+    if (!item) {
+      return undefined
+    }
+    return await this.encryptionService.decrypt(
+      item.accessToken,
+      PlaidService.encryptionContext(userId, itemId),
+    )
+  }
+
+  private static encryptionContext(userId: ModelId, itemId: string): Record<string, string> {
+    return {
+      tableName, // TODO unclear on which attacks are prevented by including tableName here.  Maybe just useful in logs?
```

Comment:
it seems logical to me that we'd want it, since we might want to encrypt something else with two other string ids, but can't think of a way an attacker would take advantage

Reply:
Yea, maybe the attack they refer to in https://aws.amazon.com/blogs/security/how-to-protect-the-integrity-of-your-encrypted-data-by-using-aws-key-management-service-and-encryptioncontext/ is the following:

1) Attacker gets access to db
2) There are 2 tables, say plaid_item and plaid_account.  Both only pass the primary key `id` as encryption context
3) Attacker moves victim's record (or just encrypted fields) from plaid_item to plaid_account, making sure plaid_account.id matches plaid_item.id and keeping plaid_account.user_id as the attacker's user id.  The attacker can then decrypt the victim's plaid_item info by requesting their own plaid_account info.

I'll keep tableName in there